ClientRuntimes/Ruby/ms-rest-azure/lib/ms_rest_azure/token_refresh_middleware.rb

```diff
@@ -0,0 +1,57 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+
+module MsRestAzure
+  #
+  # Class which handles token renewal.
+  #
+  class TokenRefreshMiddleware < Faraday::Response::Middleware
+    #
+    # Initializes a new instance of the TokenRefreshMiddleware class.
+    #
+    def initialize(app, options = nil)
+      fail ArgumentError, 'options can\'t be nil' if options.nil?
+      fail ArgumentError, 'options must contain credentials object' if options[:credentials].nil?
+      @credentials = options[:credentials]
+
+      super(app)
+    end
+
+    #
+    # Verifies whether given response is about authentication token expiration.
+    # @param response [Net::HTTPResponse] http response to verify.
+    #
+    # @return [Bool] true if response is about authentication token expiration, false otherwise.
+    def is_token_expired_response(response)
+      return false unless response.status == 401
+
+      begin
+        response_body = JSON.load(response.body)
+        error_code = response_body['error']['code']
+        error_message = response_body['error']['message']
+      rescue Exception => e
+        return false
+      end
+
+      return (error_code == 'AuthenticationFailed' && (error_message.start_with?('The access token expiry') || (error_message.start_with?('The access token is missing or invalid'))))
```

Comment:
i'm not sure how robust this is over changes in the autrhentication service. I woulkd suggest checking the token expiration (which is returned with the access token) when receiving a 401, instead of relying on message text.  You *mightr* also consider pre-emptively checking expiration before sending the request, instead of waiting for the 401